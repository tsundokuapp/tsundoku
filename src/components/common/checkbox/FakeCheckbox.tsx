import { ICheckbox } from '@/@types/Checkbox';
import { cn } from '@/helpers/twUtils';

type FakeCheckboxProps = Omit<ICheckbox, 'register' | 'name'>;

export const FakeCheckbox = ({
  label,
  description,
  checked,
  size = 'sm',
  id,
}: FakeCheckboxProps) => {
  return (
    <div className="flex items-center justify-center">
      <div className="flex h-5 items-center bg-transparent">
        <input
          id={id || 'helper-faker-checkbox'}
          aria-describedby="helper-faker-checkbox-input"
          checked={checked}
          type="checkbox"
          tabIndex={-1}
          disabled
          className={cn(
            'disabled:border-steel-400 disabled:bg-steel-400 focus-event-none checked:bg-primary focus:ring-primary dark:bg-appPrimary checked:dark:bg-primary dark:focus:ring-primary peer pointer-events-none relative mt-1 shrink-0 appearance-none rounded-sm border-2 border-[#e0e0e0] bg-white checked:border-0 focus:outline-none focus:ring-2 focus:ring-offset-0 dark:border-slate-700 checked:dark:border-0',
            {
              'h-4 w-4': size === 'sm',
              'h-5 w-5': size === 'md',
              'h-6 w-6': size === 'lg',
            },
          )}
        />
        <svg
          className={cn(
            'pointer-events-none absolute mt-1 hidden peer-checked:block',
            {
              'h-4 w-4': size === 'sm',
              'h-5 w-5': size === 'md',
              'h-6 w-6': size === 'lg',
            },
          )}
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
          fill="none"
          stroke="currentColor"
          strokeWidth="4"
          strokeLinecap="round"
          strokeLinejoin="round"
        >
          <polyline points="20 6 9 17 4 12"></polyline>
        </svg>
      </div>
      <div
        className={cn('ms-2 text-sm', {
          'text-md': size === 'md' || 'lg',
        })}
      >
        <label
          htmlFor="helper-fake-checkbox"
          className="font-medium text-gray-900 dark:text-gray-300"
        >
          {label}
        </label>
        {description && (
          <p
            id="helper-faker-checkbox-text"
            className="flex-wrap text-xs font-normal text-gray-500 dark:text-slate-500"
          >
            {description}
          </p>
        )}
      </div>
    </div>
  );
};
