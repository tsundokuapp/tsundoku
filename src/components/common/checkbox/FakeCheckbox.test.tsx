import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';

import { FakeCheckbox } from './FakeCheckbox';

describe('Testes para o componente <FakeCheckbox />', () => {
  describe('Testes de renderização', () => {
    it('Deve renderizar FakeCheckbox corretamente', () => {
      // Given
      render(
        <FakeCheckbox
          label="Checkbox Teste"
          name="checkboxTeste"
          checked={false}
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');

      // Then
      expect(checkbox).toBeInTheDocument();
    });

    it('Deve renderizar a label do FakeCheckbox corretamente', () => {
      // Given
      render(
        <FakeCheckbox
          label="Checkbox Teste"
          name="checkboxTeste"
          checked={false}
        />,
      );

      // When
      const label = screen.getByText('Checkbox Teste');

      // Then
      expect(label).toBeInTheDocument();
    });

    it('Deve renderizar o texto de apoio do FakeCheckbox corretamente', () => {
      // Given
      render(
        <FakeCheckbox
          label="Checkbox Teste"
          name="checkboxTeste"
          checked={false}
          description="Texto de apoio do Checkbox Teste"
        />,
      );

      // When
      const description = document.getElementById('helper-faker-checkbox-text');

      // Then
      expect(description).toHaveTextContent('Texto de apoio do Checkbox Teste');
    });

    it('Deve renderizar a label e o texto de apoio do FakeCheckbox simultaneamente', () => {
      // Given
      render(
        <FakeCheckbox
          label="Checkbox Teste"
          name="checkboxTeste"
          checked
          description="Texto de apoio do Checkbox Teste"
        />,
      );

      // When
      const label = screen.getByText('Checkbox Teste');
      const description = document.getElementById('helper-faker-checkbox-text');

      // Then
      expect(label).toHaveTextContent('Checkbox Teste');
      expect(description).toHaveTextContent('Texto de apoio do Checkbox Teste');
    });

    it('Deve renderizar FakeCheckbox com tamanhos diferentes baseado em props', () => {
      // Given
      render(
        <div>
          <FakeCheckbox
            label="Checkbox Teste"
            name="checkboxTeste"
            checked={false}
            size="lg"
          />
          <FakeCheckbox
            label="Checkbox Teste"
            name="checkboxTeste"
            checked={false}
            id="checkbox-default"
          />
        </div>,
      );

      // When
      const checkbox = document.getElementById('helper-faker-checkbox');
      const checkboxDefault = document.getElementById('checkbox-default');

      // Then
      expect(checkbox).toHaveClass('h-6 w-6');
      expect(checkboxDefault).toHaveClass('h-4 w-4');
    });
  });

  describe('Testes de interação e comportamento', () => {
    it('Não deve ser possível clicar no FakeCheckbox', () => {
      // Given
      render(
        <FakeCheckbox
          label="Checkbox Teste"
          name="checkboxTeste"
          checked={false}
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');
      fireEvent.click(checkbox);

      // Then
      expect(checkbox).not.toBeChecked();
    });
  });

  describe('Testes de acessibilidade', () => {
    it('Deve ter o atributo role igual a checkbox', () => {
      // Given
      render(
        <FakeCheckbox
          label="Checkbox Teste"
          name="checkboxTeste"
          checked={false}
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');

      // Then
      expect(checkbox).toBeInTheDocument();
    });

    it('Deve ter o atributo tabindex igual a -1', () => {
      // Given
      render(
        <FakeCheckbox
          label="Checkbox Teste"
          name="checkboxTeste"
          checked={false}
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');

      // Then
      expect(checkbox).toHaveAttribute('tabindex', '-1');
    });

    it('Dever ser possível focar no checkbox', () => {
      // Given
      render(
        <FakeCheckbox
          label="Checkbox Teste"
          name="checkboxTeste"
          aria-label="Checkbox de teste"
          checked={false}
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');
      checkbox.focus();

      // Then
      expect(checkbox).toHaveFocus();
    });
  });

  describe('Teste de snapshot', () => {
    it('Deve manter o snapshot do Checkbox', () => {
      // Given
      const { container } = render(
        <FakeCheckbox
          label="Checkbox Teste"
          name="checkboxTeste"
          checked={false}
        />,
      );

      // Then
      expect(container).toMatchSnapshot();
    });
  });
});
