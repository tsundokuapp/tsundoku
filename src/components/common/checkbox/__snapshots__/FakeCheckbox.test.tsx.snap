// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testes para o componente <FakeCheckbox /> Teste de snapshot Deve manter o snapshot do Checkbox 1`] = `
<div>
  <div
    class="flex items-center justify-center"
  >
    <div
      class="flex h-5 items-center bg-transparent"
    >
      <input
        aria-describedby="helper-faker-checkbox-input"
        class="disabled:border-steel-400 disabled:bg-steel-400 focus-event-none peer pointer-events-none relative mt-1 shrink-0 appearance-none rounded-sm border-2 border-[#e0e0e0] bg-white checked:border-0 checked:bg-primary focus:outline-none focus:ring-2 focus:ring-primary focus:ring-offset-0 dark:border-slate-700 dark:bg-slate-900 checked:dark:border-0 checked:dark:bg-primary dark:focus:ring-primary h-4 w-4"
        disabled=""
        id="helper-faker-checkbox"
        tabindex="-1"
        type="checkbox"
      />
      <svg
        class="pointer-events-none absolute mt-1 hidden peer-checked:block h-4 w-4"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="4"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
    <div
      class="ms-2 text-md"
    >
      <label
        class="font-medium text-gray-900 dark:text-gray-300"
        for="helper-fake-checkbox"
      >
        Checkbox Teste
      </label>
    </div>
  </div>
</div>
`;
